### PR TITLE
Hierarchical upload API optimized for folders & collections.

### PR DESCRIPTION
**Pull request by jmchilton**
This new API endpoint allows describing hierarchical data in JSON or inferring structure from archives or directories.

Datasets or archive sources can be specified via uploads, URLs, paths (if admin && allow_path_paste), library_import_dir/user_library_import_dir, and/or FTP imports. Unlike the existing library API endpoint, a mix of these on a per file source basis is allowed.

Supported "archives" include gzip, zip, bagit directories, bagit achives (with fetching and validations of downloads).

The existing upload API endpoint is quite rough to work with both in terms of adding parameters (e.g. the file type and dbkey handling in #4563 was difficult to implement, terribly hacky, and should seemingly have been trivial) and in terms of building requests (one needs to build a tool form - not describe sensible inputs in JSON). This API is built to be intelligible from an API standpoint instead of being constrained to the older style tool form. Additionally it built with hierarchical data in mind in a way that would not be easy at all enhancing the tool form components we don't even render.

This implements #5159 by allowing BagIt directories and BagIt archives (e.g. zip or tar.gz files containing BagIt direcotries) to specified as the source of items for a library folder. though a much simpler YAML description of data libraries can be translated directly into API calls into the new endpoint. To demonstrate this I've included an example script [fetch_to_library.py](https://github.com/jmchilton/galaxy/blob/upload_2.0/scripts/api/fetch_to_library.py) that implements this. The example input file included demonstrates this:

```yaml
destination:
  type: library
  name: Training Material
  description: Data for selected tutorials from https://training.galaxyproject.org.
items:
  - name: Quality Control
    description: |
      Data for sequence quality control tutorial at http://galaxyproject.github.io/training-material/topics/sequence-analysis/tutorials/quality-control/tutorial.html.
      10.5281/zenodo.61771
    items:
      - src: url
        url: https://zenodo.org/record/61771/files/GSM461178_untreat_paired_subset_1.fastq
        name: GSM461178_untreat_paired_subset_1
        ext: fastqsanger
        info: Untreated subseq of GSM461178 from 10.1186/s12864-017-3692-8
      - src: url
        url: https://zenodo.org/record/61771/files/GSM461182_untreat_single_subset.fastq
        name: GSM461182_untreat_single_subset
        ext: fastqsanger
        info: Untreated subseq of GSM461182 from 10.1186/s12864-017-3692-8
  - name: Small RNA-Seq
    description: |
      Data for small RNA-seq tutorial available at http://galaxyproject.github.io/training-material/topics/transcriptomics/tutorials/srna/tutorial.html
      10.5281/zenodo.826906
    items:
      - src: url
        url: https://zenodo.org/record/826906/files/Symp_RNAi_sRNA-seq_rep1_downsampled.fastqsanger.gz
        name: Symp RNAi sRNA Rep1
        ext: fastqsanger.gz
        info: Downsample rep1 from 10.1186/s12864-017-3692-8
      - src: url
        url: https://zenodo.org/record/826906/files/Symp_RNAi_sRNA-seq_rep2_downsampled.fastqsanger.gz
        name: Symp RNAi sRNA Rep2
        ext: fastqsanger.gz
        info: Downsample rep2 from 10.1186/s12864-017-3692-8
      - src: url
        url: https://zenodo.org/record/826906/files/Symp_RNAi_sRNA-seq_rep3_downsampled.fastqsanger.gz
        name: Symp RNAi sRNA Rep3
        ext: fastqsanger.gz
        info: Downsample rep3 from 10.1186/s12864-017-3692-8
```

This example demonstrates creating a library via this endpoint, but that destination element could be used to create an HDCA in a history or populate the contents of an existing library folder just as easily.

Once this lands in a stable release - I'll port a cleaned up version of this example script as a replacement or augmentation to the [data library script](https://github.com/galaxyproject/ephemeris/blob/master/ephemeris/setup_data_libraries.py) in Ephemeris. Unlike the script in Ephemeris and the existing library folder upload options - this handles much more metadata and allows the creation of nested folders instead of just flat lists of contents.

In future PRs I'll add filtering options to this and it will serve as the backend to #4733.

Implements #4734.

### CODE REVIEW

**Comment by bgruening** [comment]
@jmchilton this is super cool and such a yaml file turned out to be super useful in our training community. Looking forward to these nice enhancements.

What do you think about adding permissions also to this schema, so that I can specify during upload the group/role?

**Comment by martenson** [comment]
>What do you think about adding permissions also to this schema, so that I can specify during upload the group/role?

Making permissions part of the schema would make the 'bag' not portable though since it would relay on given instance's namespace. Or did you mean something like 'set access to the bag to these four users on import'  feature?

**Comment by jmchilton** [comment]
So a syntax for bag imports currently would be:

```
destination:
  type: library
  name: Training Material
  description: Data for selected tutorials from https://training.galaxyproject.org.
items_from: bagit_archive
src: url
url: http://example.org/coolbag.zip
```

or

```
destination:
  type: library
  name: Training Material
  description: Data for selected tutorials from https://training.galaxyproject.org.
items_from: bagit_archive
src: ftp_import
ftp_path: relative_ftp_path/to/coolbag.zip
```

or some other stuff for direct path imports, server_dir imports, uploads, etc... At any rate I would imagine one would stash the roles needed for library creation in the destination part:

```
destination:
  type: library
  name: Training Material
  description: Data for selected tutorials from https://training.galaxyproject.org.
  roles:
    admin:
      - name_of_role
    read: 
      - name_or_ro_role
items_from: bagit_archive
src: ftp_import
ftp_path: relative_ftp_path/to/coolbag.zip
```

Or however we would identify the different kinds of roles and the roles themselves - so I think this could make sense. The destination is pretty independent from what is being imported. Another thing is that existing library folders can be targeted currently (it is actually how I started implementing this).

```
destination:
  type: library_folder
  library_folder_id: folder_id
items_from: bagit_archive
src: ftp_import
ftp_path: relative_ftp_path/to/coolbag.zip
```

so certainly this can be used to upload bags or zips or files or directories to existing folders with custom permissions as well.

I have a TODO list for this PR that definitely has permissions on it - but it might be relegated to a follow up issue unless someone feels strongly this needs to be in the first attempt.

**Comment by martenson** [comment]
I see, so the roles here would be user emails which can be portable to some extent?

edit: because the bag does not know about groups and custom roles names/ids

**Comment by jmchilton** [comment]
> I see, so the roles here would be user emails which can be portable to some extent?

There are two things - the API and the the YAML format. They corresponding pretty directly and certainly role ids or role names in the API make a lot of sense. I wouldn't really expect for instance sample YAML files with library descriptions to be portable if they defined such roles though. Perhaps CLI parameters for settings those up or a GUI that you can paste the generic config into and then customize.

**Comment by martenson** [comment]
> GUI that you can paste the generic config into and then customize. 

That is what I was thinking. Load your bag definition and then you would have an ability to choose instance specific settings like permissions, parent folder etc. before you click the big 'Import' button.

File: lib/galaxy/workflow/modules.py [hunk 1]
```diff
@@ -809,6 +809,9 @@ def execute(self, trans, progress, invocation_step, use_cached_job=False):
         invocation = invocation_step.workflow_invocation
         step = invocation_step.workflow_step
         tool = trans.app.toolbox.get_tool(step.tool_id, tool_version=step.tool_version)
+        if tool.is_workflow_compatible:
```
Comment (bgruening) [inline lib/galaxy/workflow/modules.py]: if not?

Comment (jmchilton) [reply lib/galaxy/workflow/modules.py]: Yes - thanks!

File: lib/galaxy/tools/actions/upload_common.py [hunk 2]
```diff
@@ -334,7 +334,11 @@ def new_upload(trans, cntrller, uploaded_dataset, library_bunch=None, history=No
 def get_uploaded_datasets(trans, cntrller, params, precreated_datasets, dataset_upload_inputs, library_bunch=None, history=None):
     uploaded_datasets = []
     for dataset_upload_input in dataset_upload_inputs:
-        uploaded_datasets.extend(dataset_upload_input.get_uploaded_datasets(trans, params))
+        try:
+            uploaded_datasets.extend(dataset_upload_input.get_uploaded_datasets(trans, params))
+        except AttributeError:
+            # TODO: refine...
```
Comment (bgruening) [inline lib/galaxy/tools/actions/upload_common.py]: I guess this should be fixed before merging.

File: lib/galaxy/tools/data_fetch.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+
+
+def _fetch_target(upload_config, target):
+    destination = target.get("destination", None)
+    assert destination, "No destination defined."
+
+    def expand_elements_from(target_or_item):
+        elements_from = target_or_item.get("elements_from", None)
+        items = None
+        assert not elements_from or elements_from in ["archive", "bagit", "bagit_archive", "directory"], elements_from
```
Comment (bgruening) [inline lib/galaxy/tools/data_fetch.py]: This list should move to the top of the file imho as a global.

File: lib/galaxy/tools/parameters/output_collect.py [hunk 4]
```diff
@@ -149,6 +165,124 @@ def collect_dynamic_collections(
         inp_data,
         input_dbkey,
     )
+    log.info(tool_provided_metadata)
```
Comment (bgruening) [inline lib/galaxy/tools/parameters/output_collect.py]: Some description would be nice here, or is this a debug leftover?

File: lib/galaxy/tools/parameters/output_collect.py [hunk 5]
```diff
@@ -149,6 +165,124 @@ def collect_dynamic_collections(
...
+                            dbkey=dbkey,
+                            name=name,
+                            filename=discovered_file.path,
+                            info=info,
+                            library_folder=library_folder,
+                            link_data=link_data
+                        )
+
+            add_elements_to_folder(elements, library_folder)
+        elif destination_type == "hdca":
```
Comment (bgruening) [inline lib/galaxy/tools/parameters/output_collect.py:224]: Would be nice to add some comments here and the other destination points for people that do not know the difference between hdca/hdas.

**Review by bgruening — COMMENTED** [review]
A few more comments. So many tests, that awesome!

**Comment by jmchilton** [comment]
@bgruening thanks for the detailed review, I believe 2e22802be4422dfcedbb593c1130b86230442c31 should address all your comments.

File: lib/galaxy/tools/parameters/output_collect.py (hunk 5 above)
Comment (jmchilton) [reply lib/galaxy/tools/parameters/output_collect.py:224]: Github (at least in my browser) has decided to keep this comment visible and not below a "Show outdated" collapse for some reason but I have added this comment to the line below this in 2e22802be4422dfcedbb593c1130b86230442c31 - and I have added a big comment at the top of the if/elif chain.

**Comment by anatskiy** [comment]
@jmchilton I'm trying to upload a CSV file to the History and every single time I get an error. What am I missing? My yml file is:

```
destination:
  type: hdas
  name: Some dataset
  description: some description
items:
  - name: Dataset 1
    src: http://samplecsvs.s3.amazonaws.com/SalesJan2009.csv
```
s

**Comment by jmchilton** [comment]
Thanks for testing @anatskiy!

- How did you use that YAML file? With the script included in the PR?
- Can you click the info icon and include the full stderr - the traceback is cutoff here but I think Galaxy probably has a more complete one available?

**Comment by anatskiy** [comment]
@jmchilton I just noticed url params were wrong, it should be like this:
```
src: url
url: http://samplecsvs.s3.amazonaws.com/SalesJan2009.csv
```
The error is gone, but the file can't be processed (doesn't reach the success/fail status). The loader just spins forever...

> How did you use that YAML file? With the script included in the PR?

Exactly.

**Comment by jmchilton** [comment]
@anatskiy definitely there is something broken here - I'll take a look. I think maybe the late addition of the dataset pre-creation #5609 has some problems when creating individual datasets this way. I'll try to fix.

**Comment by jmchilton** [comment]
@anatskiy So I hit a few problems getting your example to work but I've come up with solutions and test cases:

I was right that this broke down quite a bit when I merged #5609 back into this branch, but that breakage should all be fixed with https://github.com/galaxyproject/galaxy/pull/5220/commits/6252a9e3292a3b07ea679f0be3e25014b660d113.

Then I noticed that Galaxy doesn't consistently sniff CSV files in the same way depending on how you are uploading - so I opened #5643 for the existing uploader (which is included in this PR) and added a fix for this new uploader.

Finally I added in https://github.com/galaxyproject/galaxy/pull/5220/commits/c6951128779fd1aa0cc13db292e044649b8bd66f which has a bunch more tests for the new uploader including CSV tests derived from your file.

So

```yaml
destination:
  type: hdas
  name: Some dataset
  description: some description
items:
  - name: Dataset 1
    src: url 
    url: http://samplecsvs.s3.amazonaws.com/SalesJan2009.csv
```

should now upload just fine. The one caveat I did learn is that your file has carriage returns in it - that Python doesn't by default know how to parse when sniffing as a csv, so this file ends up as a txt input. This isn't a problem for the regular uploader because it always converts carriage returns to posix newlines. This API doesn't do that by default because it is less eager to mess with your data - but it can be switched on. So to get this to sniff as CSV you can use:

```yaml
destination:
  type: hdas
  name: Some dataset
  description: some description
items:
  - name: Dataset 1
    src: url 
    url: http://samplecsvs.s3.amazonaws.com/SalesJan2009.csv
    to_posix_lines: true
```

or just

```yaml
destination:
  type: hdas
  name: Some dataset
  description: some description
items:
  - name: Dataset 1
    src: url 
    url: http://samplecsvs.s3.amazonaws.com/SalesJan2009.csv
    to_posix_lines: true
    ext: csv
```

Thanks again for trying it out, hope these fixes help.

**Comment by anatskiy** [comment]
@jmchilton thanks a lot! It works perfectly now! 

I have just one question. Is it possible to upload data with just a POST request? Or such API doesn't exist?

**Comment by bgruening** [comment]
@jmchilton if you can resolve the conflict please I think we are ready to merge.

**Comment by jmchilton** [comment]
@bgruening Awesome - thanks for the review. I've rebased and resolved the conflicts.

@anatskiy Yeah - it is possible via the API - that publication script doesn't work that way though. There are examples in the test code - https://github.com/galaxyproject/galaxy/pull/5220/files#diff-a1e38277f26f32a21c0878b6b640accbR191.

So any dictionary in the API that can support for instance ``{"src": "path", "path": "/path/to"}`` or ``{"src": "url", "url": "http://example.com"}`` can also support ``{"src": "files"}``. This means treat the multi-part upload parameters ``files_0``, ``files_1``, etc... as an iterator and grab "the next one". So the number of instances of ``{"src": "files"}`` appearing should match the number of files attached via multi-part posts.

It doesn't yet support ``{"src": "paste"}`` - so you can't put the content directly in a simple non-multi-part POST I think. It will need that at some point to serve as the main upload GUI backend, but it also doesn't yet tackle composite files which will also be needed someday. I think the structure of the JSON though will allow both of those more naturally than the older endpoint.

**Comment by bgruening** [comment]
Cool thanks a lot @jmchilton.
ping @sneumann and @pcm32 this will make you happy I suppose :)

**Pull request merged**